numpy/lib/format.py

```diff
@@ -496,6 +497,10 @@ def _read_array_header(fp, version):
     #   "descr" : dtype.descr
     header = _filter_header(header)
     try:
+        old_header = header
+        # replace paddings like ('', '|V4')
+        header = re.sub(r"\s*,*\s*\(\s*''\s*,\s*'\|*V\d'\s*\)", '', header)
+        aligned = header != old_header
```

Comment:
Hmm,  this test assumes that the pressence of padding-fields implies an aligned dtype. but actually there are many ways to create a dtype which has padding fields yet which is not "aligned".

For instance, try saving and loading the array
```
>>> a = np.zeros(3, dtype={'names': ['a', 'b'], 'formats': ['i4', 'i4'], 'offsets': [1, 7]})
>>> np.save('test', a)
>>> np.load('test.npz')
```

Also, I think we should avoid regular expression processing if possible, I am worried it is too fragile. It would be better to turn the string into a list of tuples first using save_eval, and then do some computations on the list of tuples to get the right offsets, names, dtypes.

Reply:
Is it possible to save npy where fields in dtype have empty names?
It will be more complex code to rebuild list of tuples

Reply:
Also, it technically *is* possible to create a dtype with empty field, but only using the dict specification:
```python
np.dtype({'names': [''], 'formats': 'i4'})
```
(see [this comment](https://github.com/numpy/numpy/issues/7797#issuecomment-230186999))

I think that if we want to fix up the save/load code here, we may want to disallow that possibility, to avoid confusion when such an array can't be saved (after this PR).